src/rust/engine/fs/brfs/src/main.rs

```diff
@@ -0,0 +1,941 @@
+    store
+      .record_directory(&test_directory.directory(), false)
+      .wait()
+      .expect("Storing directory");
+    store
+      .record_directory(&recursive_directory.directory(), false)
+      .wait()
+      .expect("Storing directory");
+
+    let _fs = mount(mount_dir.path(), store).expect("Mounting");
+    let virtual_dir = mount_dir
+      .path()
+      .join("directory")
+      .join(digest_to_filepath(&recursive_directory.digest()));
+    let treats = virtual_dir.join("treats");
+    assert_eq!(treat_bytes.bytes(), file::contents(&treats));
+    assert!(!file::is_executable(&treats));
+
+    let roland = virtual_dir.join("cats").join("roland");
+    assert_eq!(test_bytes.bytes(), file::contents(&roland));
+    assert!(!file::is_executable(&roland));
+  }
+
+  #[test]
+  fn files_are_correctly_executable() {
+    let store_dir = TempDir::new("store").unwrap();
+    let mount_dir = TempDir::new("mount").unwrap();
+
+    let store = fs::Store::local_only(
+      store_dir.path(),
+      Arc::new(fs::ResettablePool::new("test-pool-".to_string())),
+    ).expect("Error creating local store");
+
+    let treat_bytes = TestData::catnip();
+    let directory = TestDirectory::with_mixed_executable_files();
+
+    store
+      .store_file_bytes(treat_bytes.bytes(), false)
+      .wait()
+      .expect("Storing bytes");
+    store
+      .record_directory(&directory.directory(), false)
+      .wait()
+      .expect("Storing directory");
+
+    let _fs = mount(mount_dir.path(), store).expect("Mounting");
+    let virtual_dir = mount_dir
+      .path()
+      .join("directory")
+      .join(digest_to_filepath(&directory.digest()));
+    assert_eq!(vec!["feed", "food"], file::list_dir(&virtual_dir));
+    assert!(file::is_executable(&virtual_dir.join("feed")));
+    assert!(!file::is_executable(&virtual_dir.join("food")));
+  }
+
+  pub fn digest_to_filepath(digest: &hashing::Digest) -> String {
+    format!("{}-{}", digest.0, digest.1)
+  }
+}
+
+// TODO: Write a bunch more syscall-y tests to test that each syscall for each file/directory type
+// acts as we expect.
+#[cfg(test)]
+mod syscall_tests {
+
+  use fs;
+  use futures::Future;
+  use libc;
+  use std::sync::Arc;
+  use super::mount;
+  use super::test::digest_to_filepath;
+  use tempdir::TempDir;
+  use testutil::data::TestData;
+  use std::ffi::CString;
+  use std::path::Path;
+
+  #[test]
+  fn read_file_by_digest_exact_bytes() {
+    let store_dir = TempDir::new("store").unwrap();
+    let mount_dir = TempDir::new("mount").unwrap();
+
+    let store = fs::Store::local_only(
+      store_dir.path(),
+      Arc::new(fs::ResettablePool::new("test-pool-".to_string())),
+    ).expect("Error creating local store");
+
+    let test_bytes = TestData::roland();
+
+    store
+      .store_file_bytes(test_bytes.bytes(), false)
+      .wait()
+      .expect("Storing bytes");
+
+    let _fs = mount(mount_dir.path(), store).expect("Mounting");
+
+    unsafe {
```

Comment:
There is more under this block than really needs to be... could do just the calls between `open` and `read` under the block.

Reply:
(this is a very nice test by the way.)